### PR TITLE
[BEAM-9175] Introduce an autoformatting tool to Python SDK

### PR DESCRIPTION
**Pull request by kamilwu**
This PR adds yapf configuration and all the changes made after running it over the entire Python SDK codebase.
The result is fully compliant with PEP8 (lint check passes, the issues were manually resolved).

Some of the examples and rules of the newly enforced style (they can be put under discussion and reviewed):
* Multiple dict entries can be put in the same line. 
before:
https://github.com/apache/beam/blob/b689fffa4afcc3b655e1c831d10303b2c89b43dd/sdks/python/apache_beam/io/avroio_test.py#L72-L91
after:
https://github.com/apache/beam/blob/cfa903402e9dd0213e5d837722ce22800c4669f3/sdks/python/apache_beam/io/avroio_test.py#L71-L76
With this setting, it seems the result conforms more to the already used style (lower number of lines changed).

* Yapf tries to put the first parameter on the same line where the function begins:
https://github.com/apache/beam/blob/cfa903402e9dd0213e5d837722ce22800c4669f3/sdks/python/apache_beam/utils/annotations_test.py#L56-L61
https://github.com/apache/beam/blob/cfa903402e9dd0213e5d837722ce22800c4669f3/sdks/python/apache_beam/io/gcp/bigquery.py#L398-L407

...unless some of parameters have type annotation or are too long (then it does the split before the first argument):
https://github.com/apache/beam/blob/cfa903402e9dd0213e5d837722ce22800c4669f3/sdks/python/apache_beam/runners/direct/watermark_manager.py#L129-L136

Indent width is 2 and continuation indent is 4.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: sdks/python/apache_beam/typehints/opcodes.py [hunk 1]
```diff
@@ -268,14 +266,14 @@ def load_attr(state, arg):
   name = state.get_name(arg)
   if isinstance(o, Const) and hasattr(o.value, name):
     state.stack.append(Const(getattr(o.value, name)))
-  elif (inspect.isclass(o) and
-        isinstance(getattr(o, name, None),
-                   (types.MethodType, types.FunctionType))):
+  elif (inspect.isclass(o)
```
Comment (udim) [inline sdks/python/apache_beam/typehints/opcodes.py]: Not sure I like this change: `and` should be at the end of this line.

File: sdks/python/apache_beam/typehints/trivial_inference.py [hunk 2]
```diff
@@ -68,24 +68,23 @@ def instance_to_type(o):
...
+      ]]]
     else:
       return typehints.Set[typehints.Any]
   elif t == dict:
     if len(o) > 0:
       return typehints.Dict[
           typehints.Union[[instance_to_type(k) for k, v in o.items()]],
-          typehints.Union[[instance_to_type(v) for k, v in o.items()]],
-      ]
+          typehints.Union[[instance_to_type(v) for k, v in o.items()]], ]
```
Comment (udim) [inline sdks/python/apache_beam/typehints/trivial_inference.py]: The `, ]` at the end looks odd.

File: sdks/python/apache_beam/typehints/trivial_inference.py [hunk 3]
```diff
@@ -303,10 +302,8 @@ def infer_return_type(c, input_types, debug=False, depth=5):
     elif inspect.isclass(c):
       if c in typehints.DISALLOWED_PRIMITIVE_TYPES:
         return {
-            list: typehints.List[Any],
-            set: typehints.Set[Any],
-            tuple: typehints.Tuple[Any, ...],
-            dict: typehints.Dict[Any, Any]
+            list: typehints.List[Any], set: typehints.Set[Any], tuple:
```
Comment (udim) [inline sdks/python/apache_beam/typehints/trivial_inference.py]: This change makes things less readable IMO.

File: sdks/python/apache_beam/typehints/trivial_inference.py [hunk 4]
```diff
@@ -356,8 +353,8 @@ def infer_return_type_func(f, input_types, debug=False, depth=0):
   yields = set()
   returns = set()
   # TODO(robertwb): Default args via inspect module.
-  local_vars = list(input_types) + [typehints.Union[()]] * (len(co.co_varnames)
-                                                            - len(input_types))
+  local_vars = list(input_types) + [typehints.Union[
```
Comment (udim) [inline sdks/python/apache_beam/typehints/trivial_inference.py]: same

File: sdks/python/apache_beam/typehints/trivial_inference_test_py3.py [hunk 5]
```diff
@@ -43,8 +42,8 @@ def testBuildListUnpack(self):
 
   def testBuildTupleUnpack(self):
     # Lambda uses BUILD_TUPLE_UNPACK opcode in Python 3.
-    self.assertReturnType(typehints.Tuple[int, str, str],
-                          lambda _list1, _list2: (*_list1, *_list2, *_list2),
+    self.assertReturnType(typehints.Tuple[int, str, str], lambda _list1, _list2:
+                          (*_list1, *_list2, *_list2),
```
Comment (udim) [inline sdks/python/apache_beam/typehints/trivial_inference_test_py3.py]: ALLOW_MULTILINE_LAMBDAS = false

File: sdks/python/apache_beam/typehints/opcodes.py (hunk 1 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/opcodes.py]: I believe the knob
SPLIT_BEFORE_LOGICAL_OPERATOR = false
would change this.

File: sdks/python/apache_beam/typehints/trivial_inference.py (hunk 3 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/trivial_inference.py]: The knob for this is each_dict_entry_on_separate_line I believe.

**Comment by mwalenia** [comment]
retest this please

File: sdks/python/apache_beam/typehints/trivial_inference_test_py3.py (hunk 5 above)
Comment (kamilwu) [reply sdks/python/apache_beam/typehints/trivial_inference_test_py3.py]: Weird. Looks like this knob is turned off by default, but the lambda is getting split anyway

File: sdks/python/apache_beam/typehints/opcodes.py (hunk 1 above)
Comment (kamilwu) [reply sdks/python/apache_beam/typehints/opcodes.py]: That's right, it worked

File: sdks/python/apache_beam/typehints/trivial_inference_test_py3.py (hunk 5 above)
Comment (kamilwu) [reply sdks/python/apache_beam/typehints/trivial_inference_test_py3.py]: I've disabled formatting for this block

**Comment by chadrik** [comment]
We should add a pre-commit lint job to run `yapf`, otherwise the codebase will drift from the proper yapf formatting over time, and will be rectified all at once in some unrelated PR when a contributor finally runs it.  I think we can all agree we don't want to be reviewing unrelated style changes in a PR. 

From the docs:

> If --diff is supplied, YAPF returns zero when no changes were necessary, non-zero otherwise (including program error). You can use this in a CI workflow to test that code has been YAPF-formatted.

Once this is in place we can dismantle the pylint and flake8 checks that overlap with yapf.

**Comment by robertwb** [comment]
I would suggest COALESCE_BRACKETS , SPLIT_BEFORE_FIRST_ARGUMENT, SPLIT_ALL_TOP_LEVEL_COMMA_SEPARATED_VALUES, and possibly SPLIT_BEFORE_EXPRESSION_AFTER_OPENING_PAREN to reduce churn. 

SPLIT_ARGUMENTS_WHEN_COMMA_TERMINATED could also give authors a bit more freedom.

**Comment by robertwb** [comment]
Why was SPLIT_BEFORE_LOGICAL_OPERATOR disabled? Also, +1 to a precommit.

**Comment by udim** [comment]
> Why was SPLIT_BEFORE_LOGICAL_OPERATOR disabled? Also, +1 to a precommit.

I asked for that: https://github.com/apache/beam/pull/10684/files/1250a914122615df951619361cc7f961f51e60b9..6baa4d4d2d8cf6e14348c55a521e38e9f84b8d21#diff-e710cc10a73c533a8ef446f99dce944e

File: sdks/python/apache_beam/typehints/trivial_inference.py (hunk 2 above)
Comment (kamilwu) [reply sdks/python/apache_beam/typehints/trivial_inference.py]: It's OK now.

File: sdks/python/apache_beam/typehints/trivial_inference.py (hunk 3 above)
Comment (kamilwu) [reply sdks/python/apache_beam/typehints/trivial_inference.py]: Done.

**Comment by kamilwu** [comment]
@robertwb I managed to add all knobs you asked for. I think the code looks better now (much less weird lines, that's for sure).

I also added a pre-commit job that runs yapf with --diff option.

**Comment by kamilwu** [comment]
There is still a number of pylint issues `Wrong continued indentation`. Most of them appear because of how lambda is formatted. For example:
https://github.com/apache/beam/blob/7db61fbf2dd6eac4ffb542e684260edf0d892fea/sdks/python/apache_beam/io/gcp/bigquery_test.py#L908-L910

I don't know yet how to deal with it. Unless there is a knob for this, I'll just put a # yapf: disable comments in these places.

**Comment by Ardagan** [comment]
Is it possible to split change to one that adds formatter and one that shows what formatting actually looks like?

**Comment by Ardagan** [comment]
+1 for running yapf on pre-commit. (Can be done in separate PR)

**Comment by udim** [comment]
> There is still a number of pylint issues `Wrong continued indentation`. Most of them appear because of how lambda is formatted. For example:
> https://github.com/apache/beam/blob/7db61fbf2dd6eac4ffb542e684260edf0d892fea/sdks/python/apache_beam/io/gcp/bigquery_test.py#L908-L910
> 
> I don't know yet how to deal with it. Unless there is a knob for this, I'll just put a # yapf: disable comments in these places.

If we adopt yapf's indentation style, then we can disable the same check in pylint.

**Comment by Ardagan** [comment]
> > There is still a number of pylint issues `Wrong continued indentation`. Most of them appear because of how lambda is formatted. For example:
> > https://github.com/apache/beam/blob/7db61fbf2dd6eac4ffb542e684260edf0d892fea/sdks/python/apache_beam/io/gcp/bigquery_test.py#L908-L910
> > 
> > I don't know yet how to deal with it. Unless there is a knob for this, I'll just put a # yapf: disable comments in these places.
> 
> If we adopt yapf's indentation style, then we can disable the same check in pylint.

If we utilize autoformatter, especially on pre-commit, it would be feasible to disable all formatting checks on linter. They become obsolete.

**Comment by kamilwu** [comment]
> Is it possible to split change to one that adds formatter and one that shows what formatting actually looks like?

It's already split like this. The first commit only adds the formatter, and the second commit is the one that applies formatting changes to the code. Or do you mean creating a separate PR?

**Comment by kamilwu** [comment]
> If we adopt yapf's indentation style, then we can disable the same check in pylint.

That's true. If that's so, I can disable that particular pylint check which causes problems. The question is, are we OK with how some lambdas are formatted?

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
retest this please

**Comment by kamilwu** [comment]
R: @udim @chadrik @robertwb Please take a look once again.

**Comment by mwalenia** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by mwalenia** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run PythonFormatter PreCommit

**Comment by robertwb** [comment]
I'm not a huge fan of how lambdas are formatted (tempted to file a bug against yapf), but there aren't too many of them so I'd be fine either disabling lint or disabling yapf for those lines.

**Comment by kamilwu** [comment]
Pylint shows 67 bad-continuation warnings, but some of them refer to the same malformed lambda. This means that we have 30-40 malformed lambdas in the whole project. I agree this isn't a large number.

@robertwb I've already disable lint for those lines in this PR

**Comment by kamilwu** [comment]
`Java_Examples_Dataflow` failure is unrelated (this check is flaky)

**Comment by udim** [comment]
LGTM

**Comment by mwalenia** [comment]
Run PythonFormatter PreCommit

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
Run PythonFormatter PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by aaltay** [comment]
Is it possible to update some docs/wiki, explaining how users could run this locally?

**Comment by kamilwu** [comment]
> Is it possible to update some docs/wiki, explaining how users could run this locally?

Yes, luckily I have edit access to the wiki. I'll prepare an instruction. I think [Python Tips](https://cwiki.apache.org/confluence/display/BEAM/Python+Tips) is the best place for this.

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
Run PythonFormatter PreCommit

**Comment by kamilwu** [comment]
Thanks everyone!

I've added an appropriate instruction here: https://cwiki.apache.org/confluence/display/BEAM/Python+Tips

File: sdks/python/test-suites/tox/py37/build.gradle [hunk 6]
```diff
@@ -35,6 +35,9 @@ lint.dependsOn lintPy37
 toxTask "mypyPy37", "py37-mypy"
 lint.dependsOn mypyPy37
 
+toxTask "formatter", "py3-yapf-check"
+check.dependsOn formatter
```
Comment (udim) [inline sdks/python/test-suites/tox/py37/build.gradle]: I missed this in my review. The lint precommit runs :pythonLintPreCommit, which depends on `lint` tasks.
Suggestion:
```
lint.dependsOn formatter
```

Comment (udim) [reply sdks/python/test-suites/tox/py37/build.gradle]: Nevermind, I see there's a new job named PythonFormatter

**Pull request merged**